File: frontend/src/types/recommendation.ts
/**
 * Recommendation-related TypeScript interfaces
 */

export interface RecommendationOutput {
  clothing_items: string[]
  advice: string
  weather_warnings?: string | null
  emoji?: string | null
}

export interface WeatherSnapshot {
  city: string
  temperature: number
  weather: string
  humidity: number
  wind_direction: string
  wind_power: string
  date: string
  day_of_week: string
}

export interface RecommendationResponse {
  id: string
  user_id: number
  city: string
  weather_data: WeatherSnapshot
  clothing_items: string[]
  advice: string
  weather_warnings?: string | null
  emoji_summary?: string | null
  cached: boolean
  cost_estimate?: number | null
  created_at: string
}

export interface RecommendationCreate {
  city: string
  date?: string | null // used by single-day streaming endpoint only
  days?: 1 | 2 | 3
}

export interface RecommendationListResponse {
  items: RecommendationResponse[]
  total: number
  limit: number
  offset: number
}

// Streaming event types
export interface StreamStartEvent {
  type: 'start'
  message: string
}

export interface StreamTokenEvent {
  type: 'token'
  content: string
}

export interface StreamDataEvent {
  type: 'data'
  field: string
  value: any
}

export interface StreamErrorEvent {
  type: 'error'
  message: string
}

export interface StreamDoneEvent {
  type: 'done'
  message: string
  recommendation_id?: string | null
}

export type StreamEvent =
  | StreamStartEvent
  | StreamTokenEvent
  | StreamDataEvent
  | StreamErrorEvent
  | StreamDoneEvent

// User profile types
export interface UserProfile {
  gender?: string | null
  age?: number | null
  identity?: string | null
  style?: string | null
  temperature_sensitivity?: string | null
  activity_context?: string | null
  other_preferences?: string | null
}

export interface UserProfileUpdate extends UserProfile {}

// Admin multi-day recommendation types

export interface AdminGenerateMultiDayRequest {
  user_id: number
  city_code: string
  send_email: boolean
}

export interface DailyRecommendation {
  date: string
  date_label: string
  recommendation: RecommendationResponse
  weather_summary: string
}

export interface UserBasicInfo {
  id: number
  email: string
  full_name: string | null
}

export interface MultiDayRecommendationResponse {
  user: UserBasicInfo
  city: string
  city_code: string
  recommendations: DailyRecommendation[]
  email_sent: boolean
  generated_at: string
}

File: backend/app/schemas/recommendation.py
"""Recommendation schemas."""

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class RecommendationOutput(BaseModel):
    """Structured output from LLM for AI Recommendations."""

    clothing_items: list[str] = Field(
        ...,
        description="List of specific clothing items (e.g., ['羽绒服', '保暖裤', '围巾'])",
    )
    advice: str = Field(
        ...,
        min_length=10,
        max_length=500,
        description="Personalized advice (3-5 sentences with emojis)",
    )
    weather_warnings: list[str] | None = Field(
        default=None,
        description="Weather-related warnings (e.g., ['雾天注意安全', '建议带伞'])",
    )
    emoji_summary: str = Field(
        ...,
        max_length=50,
        description="Emoji summary of the recommendation (e.g., '🧥🧣❄️')",
    )


class WeatherSnapshot(BaseModel):
    """Weather data snapshot at recommendation time."""

    city: str
    temperature: float
    weather: str
    humidity: float
    wind_direction: str
    wind_power: str
    report_time: str


class RecommendationCreate(BaseModel):
    """Schema for creating recommendations (1-3 days)."""

    city: str = Field(..., min_length=1, description="City name or AD code")
    days: int = Field(
        default=1, ge=1, le=3, description="Number of days to generate (1-3)"
    )


class RecommendationResponse(BaseModel):
    """Response schema for recommendation."""

    id: str
    user_id: int
    city: str
    weather_data: dict[str, Any]
    clothing_items: list[str]
    advice: str
    weather_warnings: list[str] | None
    emoji_summary: str
    cached: bool = Field(
        default=False, description="Whether the recommendation is from cache"
    )
    cost_estimate: float | None
    created_at: datetime

    model_config = {"from_attributes": True}


class RecommendationListResponse(BaseModel):
    """Response schema for list of recommendations."""

    items: list[RecommendationResponse]
    total: int
    limit: int
    offset: int


# Multi-day recommendation schemas for admin use


class AdminGenerateMultiDayRequest(BaseModel):
    """Request schema for admin to generate multi-day recommendations for a user."""

    user_id: int = Field(
        ..., description="Target user ID to generate recommendations for"
    )
    city_code: str = Field(
        ..., min_length=6, max_length=6, description="6-digit city AD code"
    )
    send_email: bool = Field(
        default=False, description="Automatically send email after generation"
    )


class DailyRecommendation(BaseModel):
    """Single day recommendation in multi-day response."""

    date: str = Field(..., description="Date (YYYY-MM-DD)")
    date_label: str = Field(..., description="Chinese label (今天, 明天, 后天)")
    recommendation: RecommendationResponse = Field(
        ..., description="Full recommendation object"
    )
    weather_summary: str = Field(
        ..., description="Brief weather summary (e.g., '晴天，最高15°C，最低5°C')"
    )


class UserBasicInfo(BaseModel):
    """Basic user info for multi-day response."""

    id: int
    email: str
    full_name: str | None


class MultiDayRecommendationResponse(BaseModel):
    """Response schema for multi-day recommendations."""

    user: UserBasicInfo = Field(..., description="Target user information")
    city: str = Field(..., description="City name")
    city_code: str = Field(..., description="City AD code")
    recommendations: list[DailyRecommendation] = Field(
        ..., description="List of daily recommendations (3 items)"
    )
    email_sent: bool = Field(..., description="Whether email was sent")
    generated_at: datetime = Field(
        ..., description="Timestamp when recommendations were generated"
    )
